library/x509_crt.c

```diff
@@ -1182,40 +1177,37 @@ int mbedtls_x509_crt_parse_path( mbedtls_x509_crt *chain, const char *path )
     FindClose( hFind );
 #else /* _WIN32 */
     int t_ret;
-    int snp_ret;
-    struct stat sb;
-    struct dirent *entry;
+    int entry_offset;
     char entry_name[MBEDTLS_X509_MAX_FILE_PATH_LEN];
-    DIR *dir = opendir( path );
+    mbedtls_dir_t  dir = MBEDTLS_DIR_INVALID;
 
-    if( dir == NULL )
+    entry_offset = mbedtls_snprintf( entry_name, sizeof( entry_name ),
+            "%s/", path );
+    if( entry_offset < 0 || (size_t)entry_offset >= sizeof( entry_name ) )
+        return( MBEDTLS_ERR_X509_BUFFER_TOO_SMALL );
+
+    dir = mbedtls_opendir( path );
+
+    if( dir == MBEDTLS_DIR_INVALID )
         return( MBEDTLS_ERR_X509_FILE_IO_ERROR );
 
 #if defined(MBEDTLS_THREADING_C)
     if( ( ret = mbedtls_mutex_lock( &mbedtls_threading_readdir_mutex ) ) != 0 )
     {
-        closedir( dir );
+        mbedtls_closedir( dir );
         return( ret );
     }
 #endif /* MBEDTLS_THREADING_C */
 
-    while( ( entry = readdir( dir ) ) != NULL )
+    while( mbedtls_readdir( dir, entry_name + entry_offset,
+           sizeof( entry_name ) - entry_offset ) == 0 )
     {
-        snp_ret = mbedtls_snprintf( entry_name, sizeof entry_name,
-                                    "%s/%s", path, entry->d_name );
+        mbedtls_stat_t sb;
 
-        if( snp_ret < 0 || (size_t)snp_ret >= sizeof entry_name )
-        {
-            ret = MBEDTLS_ERR_X509_BUFFER_TOO_SMALL;
-            goto cleanup;
-        }
-        else if( stat( entry_name, &sb ) == -1 )
-        {
-            ret = MBEDTLS_ERR_X509_FILE_IO_ERROR;
-            goto cleanup;
-        }
+        if ( mbedtls_stat( entry_name, &sb ) != 0 )
+            return( MBEDTLS_ERR_X509_FILE_IO_ERROR);
```

Comment:
Why was the `cleanup` section removed? Don't we leave the directory open if we return here?